src/widgets/table.rs

```diff
@@ -602,6 +602,176 @@ mod tests {
+            selection_width: u16,
+        ) -> Vec<(u16, u16)> {
+            let table = Table::new(vec![]).widths(widths);
+
+            table.get_columns_widths(available_width, selection_width)
+        }
+
+        #[test]
+        fn length_constraint() {
+            // without selection, more than needed width
+            assert_eq!(
+                get_test(&[Constraint::Length(4), Constraint::Length(4)], 20, 0),
+                &[(0, 4), (5, 4)]
+            );
+
+            // with selection, more than needed width
+            assert_eq!(
+                get_test(&[Constraint::Length(4), Constraint::Length(4)], 20, 3),
+                &[(3, 4), (8, 4)]
+            );
+
+            // without selection, less than needed width
+            assert_eq!(
+                get_test(&[Constraint::Length(4), Constraint::Length(4)], 7, 0),
+                &[(0, 4), (5, 2)]
+            );
+
+            // with selection, less than needed width
+            assert_eq!(
+                get_test(&[Constraint::Length(4), Constraint::Length(4)], 7, 3),
+                &[(3, 4), (7, 0)]
+            );
+        }
+
+        #[test]
+        fn max_constraint() {
+            // without selection, more than needed width
+            assert_eq!(
+                get_test(&[Constraint::Max(4), Constraint::Max(4)], 20, 0),
+                &[(0, 4), (5, 4)]
+            );
+
+            // with selection, more than needed width
+            assert_eq!(
+                get_test(&[Constraint::Max(4), Constraint::Max(4)], 20, 3),
+                &[(3, 4), (8, 4)]
+            );
+
+            // without selection, less than needed width
+            assert_eq!(
+                get_test(&[Constraint::Max(4), Constraint::Max(4)], 7, 0),
+                &[(0, 4), (5, 2)]
+            );
+
+            // with selection, less than needed width
+            assert_eq!(
+                get_test(&[Constraint::Max(4), Constraint::Max(4)], 7, 3),
+                &[(3, 3), (7, 0)]
+            );
+        }
+
+        #[test]
+        fn min_constraint() {
+            // in its currently stage, the "Min" constraint does not grow to use the possible
+            // available length and enabling "expand_to_fill" will just stretch the last
+            // constraint and not split it with all available constraints
+
+            // without selection, more than needed width
+            assert_eq!(
+                get_test(&[Constraint::Min(4), Constraint::Min(4)], 20, 0),
+                &[(0, 4), (5, 4)]
+            );
+
+            // with selection, more than needed width
+            assert_eq!(
+                get_test(&[Constraint::Min(4), Constraint::Min(4)], 20, 3),
+                &[(3, 4), (8, 4)]
+            );
+
+            // without selection, less than needed width
+            assert_eq!(
+                get_test(&[Constraint::Min(4), Constraint::Min(4)], 7, 0),
+                &[(0, 4), (4, 3)]
```

Comment:
i just noticed this evaluates to `4+1+3 = 8`, which is more than the given area, thankfully #405 would fix this

Reply:
There's a followup I have waiting on #405 - which will fix the rounding of f64->u16, which should clear up the gaps and maybe some places that end up over allocating due to rounding.

Reply:
Oh, and characterization tests are fine to be wrong like this - they document the current state, and can be changed when we fix things.

Reply:
> Oh, and characterization tests are fine to be wrong like this - they document the current state, and can be changed when we fix things.

i know, i just wanted to mention this here, because it currently does not mention this via a comment, should i add a comment?

Reply:
actually, nevermind this calculation excludes the spacer (spacer is width `0`), so this initial comment is wrong

context:

```txt
[
    Rect { // selection
        x: 0,
        y: 0,
        width: 0,
        height: 1,
    },
    Rect { // constraint 1
        x: 0,
        y: 0,
        width: 4,
        height: 1,
    },
    Rect { // spacer
        x: 4,
        y: 0,
        width: 0,
        height: 1,
    },
    Rect { // constraint 2
        x: 4,
        y: 0,
        width: 3,
        height: 1,
    },
]
```

and #405 actually changes to:

```txt
[
    Rect { // selection
        x: 0,
        y: 0,
        width: 0,
        height: 1,
    },
    Rect { // constraint 1
        x: 0,
        y: 0,
        width: 4,
        height: 1,
    },
    Rect { // spacer
        x: 4,
        y: 0,
        width: 0,
        height: 1,
    },
    Rect { // constraint 2
        x: 3,
        y: 0,
        width: 4,
        height: 1,
    },
]
```
meaning full length would be `8`, which is more than the given `7`